### PR TITLE
oauth2: return error if error field is present in access token response

### LINKED ISSUE #441: oauth2 swallows token response error if HTTP code is 200
What version of Go are you using (go version)? go version go1.15.2 linux/amd64
What operating system and processor architecture are you using? Ubuntu, x86

The OAuth2 spec states [the authorization server should respond with a HTTP 400](https://tools.ietf.org/html/rfc6749#section-5.2) if there is an error in the access token response. The details of the error are conveyed in the `error`, `error_description`, and `error_uri` fields in the response.

GitHub OAuth does not follow this spec exactly and can return an error response ([examples](https://developer.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/)) with HTTP code 200. In this case, the error from GitHub is swallowed and instead we return [a generic error](https://sourcegraph.com/github.com/golang/oauth2@5d25da1a8d43b66f2898c444f899c7bcfd6a407e/-/blob/internal/token.go#L281-283). This makes it difficult to debug exactly what went wrong when GitHub returns an access token response error.

Reproduction steps:
* Configure GitHub as OAuth2 authorization provider for a web service that uses this package.
* Change the GitHub client secret so that it no longer matches.
* Attempt login via GitHub.
* Note that the error returned by this package is the generic `oauth2: server response missing access_token`, rather than the actual error returned by GitHub: [`incorrect_client_credentials`](https://developer.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/#incorrect-client-credentials).

Proposed fix: Parse the `error` field in the access token response. If it is non-empty, then treat the response as an error and return an error containing the `error`, `error_description`, and `error_uri` fields in the access token response. An implementation of this fix is here: https://github.com/sourcegraph/oauth2/commit/605770292164aba6b415b7b243a20006ac97b90d.

### PR DESCRIPTION
**Pull request by beyang**
Return an error containing the error, error_description, and error_uri when the error field is present in the access token response, even if the HTTP response code is 2xx.

Fixes #441

### CODE REVIEW

**Comment by bonifaido** [comment]
Related issue https://github.com/golang/oauth2/issues/274 (reported by me long ago), I guess this one will get fixed as well.

**Comment by bobbytables** [comment]
Came here because Slack does something similar and there's no way to know exactly what the error is. What's missing from this being merged?

File: internal/token.go [hunk 1]
```diff
@@ -292,3 +309,13 @@ type RetrieveError struct {
...
 }
+
+type TokenError struct {
+	Err              string
+	ErrorDescription string
+	ErrorURI         string
+}
+
+func (t *TokenError) Error() string {
+	return fmt.Sprintf("oauth2: error in token fetch repsonse: %s\nerror_description: %s\nerror_uri: %s", t.Err, t.ErrorDescription, t.ErrorURI)
```
Comment (rbvictor) [inline internal/token.go:320]: small type in `repsonse`

**Comment by rbvictor** [comment]
Hello maintainers, 
I would like to know if there are any plans for this PR to be merged into the main branch.

**Comment by windforce2** [comment]
Hello @codyoss,
Is there any timeline as to when this PR will be finally approved and merged?
If not, there should be some other way for dealing with OAuth2 errors coming in 200 responses.
Debugging errors in production has been much harder without being able to see those.

**Comment by hickford** [comment]
@bradfitz Can you help or assign a new reviewer? https://go-review.googlesource.com/c/oauth2/+/261220

**Comment by beyang** [comment]
Closing due to lack of activity

**Comment by bobheadxi** [comment]
Could https://github.com/golang/oauth2/commit/cfe200d5bb2f300f7e51d5259d993d035b3d11b3 have been an equivalent fix?

**Pull request closed without merging**